stack-graphs/src/json.rs

```diff
@@ -0,0 +1,494 @@
+                ser.serialize_field("type", "pop_scoped_symbol")?;
+                ser.serialize_field("symbol", &graph[node.symbol])?;
+                ser.serialize_field("is_definition", &node.is_definition)?;
+                ser
+            }
+            Node::PopSymbol(node) => {
+                let mut ser = serializer.serialize_struct("node", len + 3)?;
+                ser.serialize_field("type", "pop_symbol")?;
+                ser.serialize_field("symbol", &graph[node.symbol])?;
+                ser.serialize_field("is_definition", &node.is_definition)?;
+                ser
+            }
+            Node::PushScopedSymbol(node) => {
+                let mut ser = serializer.serialize_struct("node", len + 4)?;
+                ser.serialize_field("type", "push_symbol")?;
+                ser.serialize_field("symbol", &graph[node.symbol])?;
+                ser.serialize_field("scope", &graph[node.symbol])?;
+                ser.serialize_field("is_reference", &node.is_reference)?;
+                ser
+            }
+            Node::PushSymbol(node) => {
+                let mut ser = serializer.serialize_struct("node", len + 3)?;
+                ser.serialize_field("type", "push_scoped_symbol")?;
+                ser.serialize_field("symbol", &graph[node.symbol])?;
+                ser.serialize_field("is_reference", &node.is_reference)?;
+                ser
+            }
+            Node::Root(_node) => {
+                let mut ser = serializer.serialize_struct("node", len + 1)?;
+                ser.serialize_field("type", "root")?;
+                ser
+            }
+            Node::Scope(node) => {
+                let mut ser = serializer.serialize_struct("node", len + 2)?;
+                ser.serialize_field("type", "scope")?;
+                ser.serialize_field("is_exported", &node.is_exported)?;
+                ser
+            }
+        };
+
+        ser.serialize_field("id", &self.with(&node.id()))?;
+        if let Some(source_info) = source_info {
+            ser.serialize_field("source_info", &self.with(source_info))?;
+        }
+
+        ser.end()
+    }
+}
+
+impl Serialize for InStackGraph<'_, &NodeID> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let node_id = self.0;
+
+        let len = 1 + node_id.file.into_option().map(|_| 1).unwrap_or(0);
+        let mut ser = serializer.serialize_struct("node_id", len)?;
+        if let Some(file) = node_id.file.into_option() {
+            ser.serialize_field("file", &self.with_idx(file))?;
+        }
+        ser.serialize_field("local_id", &node_id.local_id)?;
+        ser.end()
+    }
+}
+
+impl Serialize for InStackGraph<'_, &Symbol> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let symbol = self.0;
+        serializer.serialize_str(symbol.as_str())
+    }
+}
```

Comment:
I _think_ this impl might not be needed — there isn't a way right now to get a `Symbol` instance from the public API. Adding a new symbol gives you a `Handle<Symbol>`, and indexing a graph with a symbol handle gives you a `&str`, not a `&Symbol`.  (That's why the `as_str` method was previously private.)